Introduce additional columns to the Transaction list UI

mordeth opened the pull request:
Fixes #7810

#### Changes proposed in this Pull Request
This pull request introduces additional and updates some of the existing columns on the Transactions list user interface and the CSV export when there are no more than 25 transactions. The new columns include:

- Customer currency and Deposit currency. The total fee will be in the Deposit currency.
- Add Amount in Customer Currency
- Rename the column from "Source" to "Payment Method" to align with the ID field.
- Clearly specify the timezone for all timestamps as UTC.
- Add Deposit ID

New columns added to the transactions list page
<img width="1412" alt="Screenshot 2023-11-30 at 11 16 37" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/d7a75abd-605c-4baa-b761-963342771b4a">

New columns added to the transactions list CSV export
<img width="1374" alt="Screenshot 2023-11-30 at 11 17 21" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/90d3d12f-957c-4c77-8863-43259b07c609">

New columns added as options to the Transaction list display settings
<img width="217" alt="Screenshot 2023-11-30 at 13 57 59" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/aafe482e-6619-467d-a870-d017f07daff0">

https://github.com/Automattic/woocommerce-payments/assets/8667118/c643326e-be9c-4ff6-9f8a-9781ca17e77e

#### Testing instructions
**Case 1**
- Go to `Payment` -> `Transactions`.
- Confirm that the column labels are updated, and new columns are added.
- Verify that all details of the new columns are displayed correctly.

**Case 2*
- Go to `Payment` -> `Transactions`.
- Click the three dots button on the top right corner of the table.
- Confirm that the column labels are updated and new columns are added to the list of available fields.
- Toggle the button for the new fields and verify that they are displayed correctly on the transactions list table.

**Case 3*
- Go to Payment -> Transactions.
- Confirm that you have no more than 25 transactions.
- Click the Download button on the top right corner.
- Check that the column labels are updated, and the new columns are included in the exported CSV file.
- Ensure that the details for the new fields are correctly displayed in the CSV.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mordeth committed:
Introduce extra columns to the Transaction list UI

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7813` or branch name `add/7591-missing-columns-export-csv` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/7591-missing-columns-export-csv) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 85079f54ab9dc8454da8d36c7c7cd38ea385046b
- Build time: 2023-12-14 09:20:59 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

anu-rock commented on changelog/add-7591-missing-columns-export-csv:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: add
+
+Introduce additional columns to the Transaction list UI and CSV export
```
I think we should mention the names of the new columns here.

mordeth replied:
Makes sense, updated!

anu-rock commented on client/transactions/list/index.tsx:
```diff
@@ -163,10 +166,41 @@ const getColumns = (
+		},
 		{
 			key: 'amount',
 			label: __( 'Amount', 'woocommerce-payments' ),
```
@mordeth @souravdebnath1986 I wonder if we should rename this to "Deposit Amount" like we did in the server-exported CSV.

mordeth replied:
@anu-rock I intentionally refrained from modifying the `Amount` label to avoid the column expanding too much, which could lead to other columns being displaced and result in an unattractive layout.

How about incorporating a small helper icon and describing the label in a tooltip? What do you think about that?

anu-rock replied:
@mordeth As we discussed during our last sync team chat, I am aligned with your suggestion about using a tooltip. We can check with Design before making a final call.

mordeth replied:
I've incorporated a new tooltip designed for use within the TableCard component to showcase additional descriptions for the heading labels. Kindly review the attached video and screenshots. cc @elizaan36 @anu-rock @souravdebnath1986 

https://github.com/Automattic/woocommerce-payments/assets/8667118/58709ff8-a977-4eb1-843f-bbcc8fee210d

Desktop
<img width="1479" alt="Screenshot 2023-12-11 at 16 49 50" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/a4de1708-c22f-400e-9a65-0cdcffffbc00">

Mobile
<img width="356" alt="Screenshot 2023-12-11 at 16 46 21" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/ba3adbb3-dc10-49c8-bda4-25811bb3351d">

anu-rock commented on client/transactions/list/index.tsx:
```diff
@@ -371,6 +413,14 @@ export const TransactionsList = (
 		};
+		const formatCustomerAmount = () => {
+			return {
+				value: txn.customer_amount / 100,
```
This may give an unexpected value for zero-decimal currencies. I think we created a utility function earlier to handle such cases, but I cannot find it at the moment.

mordeth replied:
I couldn't locate such a method within the utils. However, the same division logic is applied in the [`formatAmount` method](https://github.com/Automattic/woocommerce-payments/blob/124c31cc57b20241d0473372fc967a0135d7d4b0/client/transactions/list/index.tsx#L388).

anu-rock replied:
In that case, it could be a good opportunity to create a helper function to format amount and refactor `formatAmount` to use it too. See the following links for more context:

pepjwh-il-p2#comment-394
https://github.com/Automattic/woocommerce-payments-server/pull/3549

mordeth replied:
In my opinion, that change falls outside the scope of the current PR. I will create an issue to examine and update all instances where division by 100 is employed. WDYT?

mordeth replied:
A new issues has been created #7834

anu-rock replied:
Awesome, thanks! Marking this thread as resolved.

anu-rock reviewed (commented):
Looks good overall.

___

#### Review checklist

| Parameter  | Result |
| ------------- | ------------- |
| Approach  | ✅ All good |
| Implementation  | ✅ All good |
| Automated tests  | ✅ All good |
| Naming  | 👍  Almost good (left a comment) |
| Typing  | Not applicable |
| Code comments  | ✅ All good |
| Changelog  | 👍  Almost good (left a comment) |
| Testing instructions  | ✅ All good |
| Demo (video or JN site)  | ✅ All good |
| Manual testing  | ✅ Works as expected |

mordeth committed:
Update the changelog

mordeth committed:
Update the changelog

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

dpaun1985 reviewed (approved):
✅ LGTM

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

mordeth committed:
Add tooltip to transaction list column headings

mordeth committed:
Merge branch 'add/7591-missing-columns-export-csv' of github.com:Automattic/woocommerce-payments into add/7591-missing-columns-export-csv

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

mordeth committed:
Format customer amount considering zero-decimal

elizaan36 commented:
hi @mordeth, The tooltips look fine to me however I'm not sure if "Amount in deposit currency" is the most clear label.
A few questions:

- What are the other tooltip labels?
- What columns are visible by default?

There are some UI issues present in the transactions table that need to be looked at, for example the row labels and amounts are not aligned, which makes the overall table harder to read. 

@rogermattic Could you take a look at this and share any recommendations? I think you can click around using the [demo site](https://github.com/Automattic/woocommerce-payments/pull/7813#issuecomment-1833679583).

mordeth commented:
Thank you for your input, @elizaan36!

The tooltip was implemented to enhance clarity in the listing table header labels without altering them, as adjusting the labels would widen the columns and push the regular visible columns to the right. Tooltips for Amount, Fees, and Net labels have been included to indicate that these values are in Deposit Currency. The reason is confusion may arise from the newly added field 'Amount in Customer Currency,' which is hidden by default.

The default visible columns remain unaffected by this pull request. You can observe them from the screenshot.
![image](https://github.com/Automattic/woocommerce-payments/assets/8667118/22b129f2-ab54-4b1e-ab9d-02a96017e492)

> There are some UI issues present in the transactions table that need to be looked at, for example the row labels and amounts are not aligned, which makes the overall table harder to read.

I believe that the design improvements for the TableCard component are beyond the scope of this issue. I recommend creating a separate issue to address these enhancements. WDYT?

rogermattic commented:
Hey @elizaan36 thanks for tagging me here! I'll have a look at the demo site and will provide feedback. 
I like your suggestion @mordeth to create a separate issue for the table improvements. I'll do that! 

Personally, I think these tips are not that useful and possibly make it even more confusing. (That's just my first thought tho) All the amounts (amount, fees, net) are in the same currency, and all the hints say that they are in "deposit currency". 
Why do we need to mention that at all?

Do we have any merchant's feedback stating that the labels are unclear without more explanation?

mordeth commented:
>Personally, I think these tips are not that useful and possibly make it even more confusing. (That's just my first thought tho) All the amounts (amount, fees, net) are in the same currency, and all the hints say that they are in "deposit currency".
Why do we need to mention that at all?

Thanks for checking it, @rogermattic! In case when only the Amount field is displayed, there's no confusion. However, we introduced a new field, Customer Amount, which is hidden by default, and could potentially cause confusion for merchants if it becomes visible.

<img width="972" alt="Screenshot 2023-12-12 at 15 50 24" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/b4b9f198-5d71-4a3e-b340-5ec2cca48570">

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

rogermattic commented:
Thanks for the screenshot @mordeth !
It makes more sense now, though... I think 'Customer Amount' as a label is not great neither. Perhaps 'Payment' 
Amount paid' or 'Paid' would work better? Or even 'Original Amount' and we could use a tooltip just for this one (instead of others) saying something along the lines of 'currency paid by a customer'. WDYT?

mordeth commented:
>It makes more sense now, though... I think 'Customer Amount' as a label is not great neither. Perhaps 'Payment'
Amount paid' or 'Paid' would work better? Or even 'Original Amount' and we could use a tooltip just for this one (instead of others) saying something along the lines of 'currency paid by a customer'. WDYT?

We've already changed the name of the 'Source' field to 'Payment', that value is already in use.

@anu-rock @souravdebnath1986 What do you think about renaming the `Customer Amount` to `Amount Paid`?

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

anu-rock commented:
@mordeth 

> What do you think about renaming the Customer Amount to Amount Paid?

"Amount Paid" sounds good to me.

souravdebnath1986 commented:
Thanks all for taking a look and apologies for joining in the conversation late. 
I think `Paid Amount` makes more sense and in the same vein `Customer Currency` -> can be modified to -> `Paid Currency`. 
I would question the use of tooltips to show the meaning of the field at this stage, we don't have any feedback to say that the field names are confusing. Until then, I don't think this is necessary.  cc: @rogermattic @mordeth

rogermattic commented:
+1 on removing the tooltips as @souravdebnath1986 suggested

Just one slight doubt on the 'Paid Currency' label  – wouldn't it suggest that the column just includes currency, and not the amount paid? 
Also, just to clarify, I was suggesting changing the 'Customer Currency' -> 'Amount Paid' :)

mordeth commented:
>Also, just to clarify, I was suggesting changing the 'Customer Currency' -> 'Amount Paid' :)

We currently have two concealed columns, namely 'Customer Currency' and 'Customer Amount.' I assume we should rename them as 'Paid Currency' and 'Paid Amount,' is that correct?

Moreover, I will remove the tooltips, and if necessary in the future, we can reintroduce them.

mordeth committed:
Remove tooltip to transaction list column headings

mordeth committed:
Update customer currency & customer amount labels

mordeth commented:
I have removed the tooltips from the labels and updated the labels.
<img width="1173" alt="Screenshot 2023-12-13 at 22 23 11" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/41f1bb30-0034-4be6-9017-d94e1d30200a">

@anu-rock Could you please conduct a final review? I aim to deploy the pull request as soon as possible.

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv

anu-rock reviewed (approved):
LGTM after latest revisions 🎉

mordeth committed:
Merge branch 'develop' into add/7591-missing-columns-export-csv
